### PR TITLE
Make getPreferredExtensions pure virtual func

### PR DESCRIPTION
**Pull request by MellowYarker**
Prior to this commit, `WebSocket::getPreferredExtensions()` had a default implementation, which meant derived classes would not be required to implement the method themselves.

Derived classes probably shouldn't refer to some default implementation, so we require it to be implemented going forward.

### CODE REVIEW

File: c++/src/kj/compat/http.c++ [hunk 1]
```diff
@@ -4084,6 +4099,11 @@ private:
       KJ_FAIL_ASSERT("Bytes are not counted for the individual states of WebSocketPipeImpl.");
     }
 
+    kj::Maybe<kj::String> getPreferredExtensions(ExtensionsContext ctx) override {
+      KJ_FAIL_ASSERT(
+          "BlockedReceive WebSocket has no websocket to get a compression preference from.");
```
Comment (MellowYarker) [inline c++/src/kj/compat/http.c++:4104]: I wonder if in this case we should forward the call to the `pipe`, but I was under the impression we would have called _from_ the pipe into the `BlockedReceive` websocket, so calling back into the `pipe` would be circular.

File: c++/src/kj/compat/http.c++ [hunk 2]
```diff
@@ -4403,19 +4442,45 @@ public:
...
+    return in->pumpTo(other).attach(kj::mv(deferredUnregister));
   }
 
   uint64_t sentByteCount() override { return out->sentByteCount(); }
   uint64_t receivedByteCount() override { return in->sentByteCount(); }
 
+  kj::Maybe<kj::String> getPreferredExtensions(ExtensionsContext ctx) override {
+    // If our destinationPumpingTo doesn't give us preferred extensions, we should check our
+    // destinationPumpingFrom next.
+    KJ_IF_SOME(ws, out->destinationPumpingTo) {
```
Comment (MellowYarker) [inline c++/src/kj/compat/http.c++:4471]: It's worth a comment explaining that we read from `out` and set it on `in` when calling `pumpTo()/tryPumpFrom()` by convention to refer to the same pipe. Will add that in the relevant spots tomorrow.

File: c++/src/kj/compat/http.c++ [hunk 3]
```diff
@@ -4312,6 +4342,12 @@ public:
   uint64_t sentByteCount() override { return out->sentByteCount(); }
   uint64_t receivedByteCount() override { return in->sentByteCount(); }
 
+  kj::Maybe<kj::String> getPreferredExtensions(ExtensionsContext ctx) override {
+    // TODO(now): Do we want in or out?
```
[the remark this answers is not in the document]
Comment (MellowYarker) [reply c++/src/kj/compat/http.c++]: Thanks for writing this, it was very helpful. Also, sorry for not getting back to this until now. I had briefly skimmed over it during holidays and figured this would be a lot more work, so when other stuff came up I avoided the context switch... Should've just read this properly :man_facepalming:

File: c++/src/kj/compat/http.c++ [hunk 4]
```diff
@@ -3901,6 +3901,23 @@ public:
...
+  kj::Maybe<WebSocket&> destinationPumpingTo;
+  kj::Maybe<WebSocket&> destinationPumpingFrom;
+  // Tracks the outstanding pumpTo() and tryPumpFrom() calls currently running on the
+  // WebSocketPipeEnd, which is the destination side of this WebSocketPipeImpl. This is used by
+  // the source end to implement getPreferredExtensions().
+  //
+  // getPreferredExtensions() does not fit into the model used by all the other methods because it
+  // is not directional (not a read nor a write call).
+
+private:
```
Comment (kentonv) [inline c++/src/kj/compat/http.c++:3920]: Duplicated `private:`.

```suggestion
```

File: c++/src/kj/compat/http.c++ [hunk 5]
```diff
@@ -3901,6 +3901,23 @@ public:
     return transferredBytes;
   }
 
+  kj::Maybe<kj::String> getPreferredExtensions(ExtensionsContext ctx) override {
```
Comment (kentonv) [inline c++/src/kj/compat/http.c++:3904]: I believe this method will never be called, since the wrapper will always use the `destinationPumping{To,From}` directly. So the body of this can be `KJ_UNREACHABLE`. Same goes for all the specific state types I think.

File: c++/src/kj/compat/http.c++ [hunk 6]
```diff
@@ -4403,19 +4442,45 @@ public:
...
+    in->destinationPumpingTo = other;
+    auto deferredUnregister = kj::defer([this]() { in->destinationPumpingTo = kj::none; });
+    return in->pumpTo(other).attach(kj::mv(deferredUnregister));
   }
 
   uint64_t sentByteCount() override { return out->sentByteCount(); }
   uint64_t receivedByteCount() override { return in->sentByteCount(); }
 
+  kj::Maybe<kj::String> getPreferredExtensions(ExtensionsContext ctx) override {
+    // If our destinationPumpingTo doesn't give us preferred extensions, we should check our
```
Comment (kentonv) [inline c++/src/kj/compat/http.c++:4469]: I'd rephrase this comment like:

> We want to forward this call to whatever WebSocket the other end of the pipe is pumping to/from, if any. We'll check them in an arbitrary order and take the first one we see. But really, the hope is that both `destinationPumpingTo` and `destinationPumpingFrom` are in fact the same object. If they aren't the same, then it's not really clear whose extensions we should prefer; the choice here is arbitrary.

**Review by kentonv — APPROVED** [review]
